File: agent/internal/hub/store.go
package hub

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/Eswink/ai-control-hud/agent/internal/domain"
	"github.com/Eswink/ai-control-hud/agent/internal/events"
	_ "modernc.org/sqlite"
)

type Store struct {
	mu sync.Mutex
	db *sql.DB
}

func OpenStore(path string) (*Store, error) {
	if path == "" {
		return nil, errors.New("hub database path is required")
	}
	if path != ":memory:" {
		if err := os.MkdirAll(filepath.Dir(path), 0o750); err != nil {
			return nil, fmt.Errorf("create hub database directory: %w", err)
		}
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, fmt.Errorf("open hub database: %w", err)
	}
	db.SetMaxOpenConns(1)
	for _, statement := range []string{
		"PRAGMA foreign_keys = ON",
		"PRAGMA busy_timeout = 5000",
	} {
		if _, err := db.Exec(statement); err != nil {
			_ = db.Close()
			return nil, fmt.Errorf("configure hub database: %w", err)
		}
	}
	if path != ":memory:" {
		if _, err := db.Exec("PRAGMA journal_mode = WAL"); err != nil {
			_ = db.Close()
			return nil, fmt.Errorf("configure hub WAL: %w", err)
		}
	}
	if _, err := db.Exec(`
		CREATE TABLE IF NOT EXISTS agents (
			agent_id TEXT PRIMARY KEY,
			agent_version TEXT,
			last_seen_at TEXT NOT NULL,
			last_sent_at TEXT NOT NULL
		);

		CREATE TABLE IF NOT EXISTS snapshots (
			agent_id TEXT PRIMARY KEY,
			received_at TEXT NOT NULL,
			sent_at TEXT NOT NULL,
			state_json TEXT NOT NULL,
			FOREIGN KEY(agent_id) REFERENCES agents(agent_id) ON DELETE CASCADE
		);

		CREATE TABLE IF NOT EXISTS events (
			seq INTEGER PRIMARY KEY AUTOINCREMENT,
			event_id TEXT NOT NULL UNIQUE,
			agent_id TEXT NOT NULL,
			event_type TEXT NOT NULL,
			occurred_at TEXT NOT NULL,
			received_at TEXT NOT NULL,
			event_json TEXT NOT NULL,
			FOREIGN KEY(agent_id) REFERENCES agents(agent_id) ON DELETE CASCADE
		);

		CREATE INDEX IF NOT EXISTS idx_events_agent_seq ON events(agent_id, seq);
	`); err != nil {
		_ = db.Close()
		return nil, fmt.Errorf("initialize hub database: %w", err)
	}
	return &Store{db: db}, nil
}

func (s *Store) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.db.Close()
}

func (s *Store) RecordState(ctx context.Context, agentID string, sentAt time.Time, state domain.HudState, receivedAt time.Time) error {
	payload, err := json.Marshal(state)
	if err != nil {
		return fmt.Errorf("encode hub state: %w", err)
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("begin state transaction: %w", err)
	}
	defer func() { _ = tx.Rollback() }()
	if err := upsertAgent(ctx, tx, agentID, sentAt, receivedAt, nil); err != nil {
		return err
	}
	if _, err := tx.ExecContext(ctx, `
		INSERT INTO snapshots(agent_id, received_at, sent_at, state_json)
		VALUES (?, ?, ?, ?)
		ON CONFLICT(agent_id) DO UPDATE SET
			received_at = excluded.received_at,
			sent_at = excluded.sent_at,
			state_json = excluded.state_json
	`, agentID, formatTime(receivedAt), formatTime(sentAt), string(payload)); err != nil {
		return fmt.Errorf("persist hub snapshot: %w", err)
	}
	if err := tx.Commit(); err != nil {
		return fmt.Errorf("commit state transaction: %w", err)
	}
	return nil
}

func (s *Store) RecordHeartbeat(ctx context.Context, agentID string, sentAt, receivedAt time.Time, agentVersion *string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := upsertAgent(ctx, s.db, agentID, sentAt, receivedAt, agentVersion); err != nil {
		return err
	}
	return nil
}

func (s *Store) RecordEvents(ctx context.Context, agentID string, sentAt time.Time, batch []events.Event, receivedAt time.Time) (int, int, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return 0, 0, fmt.Errorf("begin event transaction: %w", err)
	}
	defer func() { _ = tx.Rollback() }()
	if err := upsertAgent(ctx, tx, agentID, sentAt, receivedAt, nil); err != nil {
		return 0, 0, err
	}
	accepted := 0
	duplicates := 0
	for _, event := range batch {
		payload, err := json.Marshal(event)
		if err != nil {
			return 0, 0, fmt.Errorf("encode hub event: %w", err)
		}
		result, err := tx.ExecContext(ctx, `
			INSERT OR IGNORE INTO events(
				event_id, agent_id, event_type, occurred_at, received_at, event_json
			) VALUES (?, ?, ?, ?, ?, ?)
		`, event.EventID, agentID, event.Type, formatTime(event.OccurredAt), formatTime(receivedAt), string(payload))
		if err != nil {
			return 0, 0, fmt.Errorf("persist hub event: %w", err)
		}
		rows, err := result.RowsAffected()
		if err != nil {
			return 0, 0, fmt.Errorf("read hub event insert result: %w", err)
		}
		if rows == 1 {
			accepted++
		} else {
			duplicates++
		}
	}
	if err := tx.Commit(); err != nil {
		return 0, 0, fmt.Errorf("commit event transaction: %w", err)
	}
	return accepted, duplicates, nil
}

func (s *Store) LoadState(ctx context.Context, agentID string) (domain.HudState, time.Time, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	var payload string
	var lastSeenRaw string
	err := s.db.QueryRowContext(ctx, `
		SELECT snapshots.state_json, agents.last_seen_at
		FROM snapshots
		JOIN agents ON agents.agent_id = snapshots.agent_id
		WHERE snapshots.agent_id = ?
	`, agentID).Scan(&payload, &lastSeenRaw)
	if errors.Is(err, sql.ErrNoRows) {
		return domain.HudState{}, time.Time{}, false, nil
	}
	if err != nil {
		return domain.HudState{}, time.Time{}, false, fmt.Errorf("load hub snapshot: %w", err)
	}
	var state domain.HudState
	if err := json.Unmarshal([]byte(payload), &state); err != nil {
		return domain.HudState{}, time.Time{}, false, fmt.Errorf("decode hub snapshot: %w", err)
	}
	if err := state.Validate(); err != nil {
		return domain.HudState{}, time.Time{}, false, fmt.Errorf("stored hub snapshot is invalid: %w", err)
	}
	lastSeen, err := time.Parse(time.RFC3339Nano, lastSeenRaw)
	if err != nil {
		return domain.HudState{}, time.Time{}, false, fmt.Errorf("decode hub last-seen time: %w", err)
	}
	return state, lastSeen, true, nil
}

func (s *Store) ListEvents(ctx context.Context, agentID string, after int64, limit int) ([]HubEvent, int64, error) {
	if after < 0 {
		return nil, 0, errors.New("event cursor must be non-negative")
	}
	if limit < 1 || limit > 100 {
		return nil, 0, errors.New("event page limit must be within 1..100")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	rows, err := s.db.QueryContext(ctx, `
		SELECT seq, agent_id, received_at, event_json
		FROM events
		WHERE agent_id = ? AND seq > ?
		ORDER BY seq ASC
		LIMIT ?
	`, agentID, after, limit)
	if err != nil {
		return nil, 0, fmt.Errorf("query hub events: %w", err)
	}
	defer rows.Close()
	result := make([]HubEvent, 0, limit)
	for rows.Next() {
		var seq int64
		var rowAgentID, receivedRaw, payload string
		if err := rows.Scan(&seq, &rowAgentID, &receivedRaw, &payload); err != nil {
			return nil, 0, fmt.Errorf("scan hub event: %w", err)
		}
		var event events.Event
		if err := json.Unmarshal([]byte(payload), &event); err != nil {
			return nil, 0, fmt.Errorf("decode hub event: %w", err)
		}
		if err := validateEvent(event); err != nil {
			return nil, 0, fmt.Errorf("stored hub event is invalid: %w", err)
		}
		receivedAt, err := time.Parse(time.RFC3339Nano, receivedRaw)
		if err != nil {
			return nil, 0, fmt.Errorf("decode hub event received time: %w", err)
		}
		result = append(result, HubEvent{Seq: seq, AgentID: rowAgentID, ReceivedAt: receivedAt, Event: event})
	}
	if err := rows.Err(); err != nil {
		return nil, 0, fmt.Errorf("iterate hub events: %w", err)
	}
	var latest int64
	if err := s.db.QueryRowContext(ctx, `SELECT COALESCE(MAX(seq), 0) FROM events WHERE agent_id = ?`, agentID).Scan(&latest); err != nil {
		return nil, 0, fmt.Errorf("query hub event high-water mark: %w", err)
	}
	return result, latest, nil
}

type sqlExecutor interface {
	ExecContext(context.Context, string, ...any) (sql.Result, error)
}

func upsertAgent(ctx context.Context, exec sqlExecutor, agentID string, sentAt, receivedAt time.Time, agentVersion *string) error {
	var version any
	if agentVersion != nil {
		version = *agentVersion
	}
	if _, err := exec.ExecContext(ctx, `
		INSERT INTO agents(agent_id, agent_version, last_seen_at, last_sent_at)
		VALUES (?, ?, ?, ?)
		ON CONFLICT(agent_id) DO UPDATE SET
			agent_version = COALESCE(excluded.agent_version, agents.agent_version),
			last_seen_at = excluded.last_seen_at,
			last_sent_at = excluded.last_sent_at
	`, agentID, version, formatTime(receivedAt), formatTime(sentAt)); err != nil {
		return fmt.Errorf("persist hub agent heartbeat: %w", err)
	}
	return nil
}

func formatTime(value time.Time) string {
	return value.UTC().Format(time.RFC3339Nano)
}

File: server/hub/store.py
from __future__ import annotations

import sqlite3
import threading
from datetime import datetime
from pathlib import Path

from server.hud.models import HudState

from .models import AgentEvent, HubEvent


class HubStore:
    def __init__(self, database_path: Path) -> None:
        self.database_path = database_path
        self.database_path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = threading.RLock()
        self._connection = sqlite3.connect(
            str(database_path),
            check_same_thread=False,
            isolation_level=None,
        )
        self._connection.row_factory = sqlite3.Row
        with self._lock:
            self._connection.execute("PRAGMA foreign_keys = ON")
            if str(database_path) != ":memory:":
                self._connection.execute("PRAGMA journal_mode = WAL")
            self._connection.executescript(
                """
                CREATE TABLE IF NOT EXISTS agents (
                    agent_id TEXT PRIMARY KEY,
                    agent_version TEXT,
                    last_seen_at TEXT NOT NULL,
                    last_sent_at TEXT NOT NULL
                );

                CREATE TABLE IF NOT EXISTS snapshots (
                    agent_id TEXT PRIMARY KEY,
                    received_at TEXT NOT NULL,
                    sent_at TEXT NOT NULL,
                    state_json TEXT NOT NULL,
                    FOREIGN KEY(agent_id) REFERENCES agents(agent_id) ON DELETE CASCADE
                );

                CREATE TABLE IF NOT EXISTS events (
                    seq INTEGER PRIMARY KEY AUTOINCREMENT,
                    event_id TEXT NOT NULL UNIQUE,
                    agent_id TEXT NOT NULL,
                    event_type TEXT NOT NULL,
                    occurred_at TEXT NOT NULL,
                    received_at TEXT NOT NULL,
                    event_json TEXT NOT NULL,
                    FOREIGN KEY(agent_id) REFERENCES agents(agent_id) ON DELETE CASCADE
                );

                CREATE INDEX IF NOT EXISTS idx_events_agent_seq
                    ON events(agent_id, seq);
                """
            )

    def close(self) -> None:
        with self._lock:
            self._connection.close()

    def record_state(
        self,
        agent_id: str,
        sent_at: datetime,
        state: HudState,
        received_at: datetime,
    ) -> None:
        payload = state.model_dump_json(by_alias=True)
        with self._lock:
            self._connection.execute("BEGIN IMMEDIATE")
            try:
                self._upsert_agent(
                    agent_id=agent_id,
                    sent_at=sent_at,
                    received_at=received_at,
                    agent_version=None,
                )
                self._connection.execute(
                    """
                    INSERT INTO snapshots(agent_id, received_at, sent_at, state_json)
                    VALUES (?, ?, ?, ?)
                    ON CONFLICT(agent_id) DO UPDATE SET
                        received_at = excluded.received_at,
                        sent_at = excluded.sent_at,
                        state_json = excluded.state_json
                    """,
                    (
                        agent_id,
                        received_at.isoformat(),
                        sent_at.isoformat(),
                        payload,
                    ),
                )
                self._connection.execute("COMMIT")
            except Exception:
                self._connection.execute("ROLLBACK")
                raise

    def record_heartbeat(
        self,
        agent_id: str,
        sent_at: datetime,
        received_at: datetime,
        agent_version: str | None,
    ) -> None:
        with self._lock:
            self._upsert_agent(
                agent_id=agent_id,
                sent_at=sent_at,
                received_at=received_at,
                agent_version=agent_version,
            )

    def record_events(
        self,
        agent_id: str,
        sent_at: datetime,
        events: list[AgentEvent],
        received_at: datetime,
    ) -> tuple[int, int]:
        accepted = 0
        duplicates = 0
        with self._lock:
            self._connection.execute("BEGIN IMMEDIATE")
            try:
                # Any authenticated event delivery also proves that the agent was alive at
                # received_at. A normal heartbeat still provides the steady-state signal.
                self._upsert_agent(
                    agent_id=agent_id,
                    sent_at=sent_at,
                    received_at=received_at,
                    agent_version=None,
                )
                for event in events:
                    cursor = self._connection.execute(
                        """
                        INSERT OR IGNORE INTO events(
                            event_id,
                            agent_id,
                            event_type,
                            occurred_at,
                            received_at,
                            event_json
                        ) VALUES (?, ?, ?, ?, ?, ?)
                        """,
                        (
                            event.event_id,
                            agent_id,
                            event.type,
                            event.occurred_at.isoformat(),
                            received_at.isoformat(),
                            event.model_dump_json(by_alias=True),
                        ),
                    )
                    if cursor.rowcount == 1:
                        accepted += 1
                    else:
                        duplicates += 1
                self._connection.execute("COMMIT")
            except Exception:
                self._connection.execute("ROLLBACK")
                raise
        return accepted, duplicates

    def load_state(self, agent_id: str) -> tuple[HudState, datetime] | None:
        with self._lock:
            row = self._connection.execute(
                """
                SELECT snapshots.state_json, agents.last_seen_at
                FROM snapshots
                JOIN agents ON agents.agent_id = snapshots.agent_id
                WHERE snapshots.agent_id = ?
                """,
                (agent_id,),
            ).fetchone()
        if row is None:
            return None
        state = HudState.model_validate_json(row["state_json"])
        last_seen_at = datetime.fromisoformat(row["last_seen_at"])
        return state, last_seen_at

    def list_events(self, agent_id: str, after: int, limit: int) -> list[HubEvent]:
        with self._lock:
            rows = self._connection.execute(
                """
                SELECT seq, agent_id, received_at, event_json
                FROM events
                WHERE agent_id = ? AND seq > ?
                ORDER BY seq ASC
                LIMIT ?
                """,
                (agent_id, after, limit),
            ).fetchall()

        result: list[HubEvent] = []
        for row in rows:
            event = AgentEvent.model_validate_json(row["event_json"])
            result.append(
                HubEvent(
                    **event.model_dump(),
                    seq=row["seq"],
                    agent_id=row["agent_id"],
                    received_at=datetime.fromisoformat(row["received_at"]),
                )
            )
        return result

    def _upsert_agent(
        self,
        *,
        agent_id: str,
        sent_at: datetime,
        received_at: datetime,
        agent_version: str | None,
    ) -> None:
        self._connection.execute(
            """
            INSERT INTO agents(agent_id, agent_version, last_seen_at, last_sent_at)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(agent_id) DO UPDATE SET
                agent_version = COALESCE(excluded.agent_version, agents.agent_version),
                last_seen_at = excluded.last_seen_at,
                last_sent_at = excluded.last_sent_at
            """,
            (
                agent_id,
                agent_version,
                received_at.isoformat(),
                sent_at.isoformat(),
            ),
        )
